File: frontend/lib/api/donations.ts
import { fetchApi } from './client';

export interface Donation {
  id: string;
  donor_id: string;
  food_type: string;
  quantity_kg: number;
  available_from?: string;
  available_until?: string;
  requires_refrigeration: boolean;
  created_at: string;
}

export interface DonationCreate {
  id?: string;
  donor_id: string;
  food_type: string;
  quantity_kg: number;
  available_from?: string;
  available_until?: string;
  requires_refrigeration?: boolean;
}

export const donationsApi = {
  list: () => fetchApi<Donation[]>('/api/v1/donations'),
  get: (id: string) => fetchApi<Donation>(`/api/v1/donations/${id}`),
  create: (data: DonationCreate) => fetchApi<Donation>('/api/v1/donations', {
    method: 'POST',
    body: JSON.stringify(data),
  }),
};

File: backend/db/repositories/donations.py
"""Donation repository."""

from typing import List, Optional
from sqlalchemy.orm import Session
from backend.db.models import DonationModel
from backend.api.schemas.database import DonationCreate


def create_donation(db: Session, donation: DonationCreate) -> DonationModel:
    db_donation = DonationModel(
        donor_id=donation.donor_id,
        food_type=donation.food_type,
        quantity_kg=donation.quantity_kg,
        available_from=donation.available_from,
        available_until=donation.available_until,
        requires_refrigeration=donation.requires_refrigeration,
    )
    db.add(db_donation)
    db.commit()
    db.refresh(db_donation)
    return db_donation


def get_donation(db: Session, donation_id: str) -> Optional[DonationModel]:
    return db.query(DonationModel).filter(DonationModel.id == donation_id).first()


def list_donations(db: Session, skip: int = 0, limit: int = 100) -> List[DonationModel]:
    return db.query(DonationModel).offset(skip).limit(limit).all()


class DonationRepository:
    def __init__(self, db: Session):
        self.db = db

    def create(self, donation: DonationCreate) -> DonationModel:
        return create_donation(self.db, donation)

    def get_by_id(self, donation_id: str) -> Optional[DonationModel]:
        return get_donation(self.db, donation_id)

    def list_all(self, skip: int = 0, limit: int = 100) -> List[DonationModel]:
        return list_donations(self.db, skip=skip, limit=limit)
